From f90318cf8462e5f328881b7aeaf4ec76e5a03477 Mon Sep 17 00:00:00 2001
From: Xuefer <xuefer@gmail.com>
Date: Mon, 23 Mar 2015 16:40:30 +0800
Subject: [PATCH 4/6] MI3TD device node name hack for init

MI3 TD requires some hack for partition names which seems like a migration

Change-Id: Idf251c45ee6029dc2c9bb4e8e2e8033c2c722025
---
 init/Android.mk          |  4 ++++
 init/devices.c           | 54 ++++++++++++++++++++++++++++++++++++++++++++++--
 init/init.c              |  2 ++
 liblog/logd_write.c      |  6 +++++-
 liblog/logd_write_kern.c |  6 +++++-
 5 files changed, 68 insertions(+), 4 deletions(-)

diff --git a/init/Android.mk b/init/Android.mk
index c6863bf..73ab7e8 100755
--- a/init/Android.mk
+++ b/init/Android.mk
@@ -25,6 +25,10 @@ LOCAL_SRC_FILES += bootchart.c
 LOCAL_CFLAGS    += -DBOOTCHART=1
 endif
 
+ifneq ($(strip $(INIT_MI3TD_HACK)),)
+LOCAL_CFLAGS    += -DINIT_MI3TD_HACK
+endif
+
 ifneq (,$(filter userdebug eng,$(TARGET_BUILD_VARIANT)))
 LOCAL_CFLAGS += -DALLOW_LOCAL_PROP_OVERRIDE=1 -DALLOW_DISABLE_SELINUX=1
 endif
diff --git a/init/devices.c b/init/devices.c
index 05f0aef..39a2bb1 100644
--- a/init/devices.c
+++ b/init/devices.c
@@ -57,6 +57,38 @@
 #define FIRMWARE_DIR3   "/firmware/image"
 #define DEVICES_BASE    "/devices/soc.0"
 
+#ifdef INIT_MI3TD_HACK
+static struct {
+    char from_name[64];
+    char to_name[64];
+} mi3td_block_symlinks [] = {
+    { "LNX",        "boot"        },
+    { "boot",       "LNX"         },
+    { "APP",        "system"      },
+    { "system",     "APP"         },
+    { "CAC",        "cache"       },
+    { "cache",      "CAC"         },
+    { "UDA",        "userdata"    },
+    { "userdata",   "UDA"         },
+    { "MSC",        "misc"        },
+    { "misc",       "MSC"         },
+    { "SOS",        "recovery"    },
+    { "recovery",   "SOS"         },
+    { "USP",        "staging"     },
+    { "staging",    "USP"         },
+    { "LN1",        "boot1"       },
+    { "LOG",        "logo"        },
+    { "PST",        "persist"     },
+    { "AP1",        "system1"     },
+    { "NVC",        "microboot"   },
+    { "MBU",        "microboot1"  },
+    { "EBT",        "aboot"       },
+    { "BLU",        "aboot1"      },
+    { "STO",        "storage"     },
+    { "",           ""            }
+};
+#endif
+
 extern struct selabel_handle *sehandle;
 extern char bootdevice[32];
 
@@ -607,10 +639,10 @@ static char **get_block_device_symlinks(struct uevent *uevent)
         return NULL;
     }
 
-    char **links = malloc(sizeof(char *) * 6);
+    char **links = malloc(sizeof(char *) * 7);
     if (!links)
         return NULL;
-    memset(links, 0, sizeof(char *) * 6);
+    memset(links, 0, sizeof(char *) * 7);
 
     INFO("found %s device %s\n", type, device);
 
@@ -656,6 +688,24 @@ static char **get_block_device_symlinks(struct uevent *uevent)
         else
             links[link_num] = NULL;
 
+#ifdef INIT_MI3TD_HACK
+        {
+            int i;
+
+            for (i = 0; mi3td_block_symlinks[i].from_name[0]; ++i) {
+                if (!strncmp(mi3td_block_symlinks[i].from_name, uevent->partition_name, sizeof(mi3td_block_symlinks[i].from_name))) {
+                    if (asprintf(&links[link_num], "%s/by-name/%s", link_path, mi3td_block_symlinks[i].to_name) > 0) {
+                        link_num++;
+                    }
+                    else {
+                        links[link_num] = NULL;
+                    }
+                    break;
+                }
+            }
+        }
+#endif
+
         if (is_bootdevice >= 0) {
             if (asprintf(&links[link_num], "/dev/block/bootdevice/by-name/%s", p) > 0)
                 link_num++;
diff --git a/init/init.c b/init/init.c
index 9a2a10d..6cb37eb 100644
--- a/init/init.c
+++ b/init/init.c
@@ -753,6 +753,8 @@ static void import_kernel_nv(char *name, int for_emulator)
         cnt = snprintf(prop, sizeof(prop), "ro.boot.%s", boot_prop_name);
         if (cnt < PROP_NAME_MAX)
             property_set(prop, value);
+    } else if (!strcmp(name, "syspart")) {
+        property_set("ro.syspart", value);
     }
 }
 
diff --git a/liblog/logd_write.c b/liblog/logd_write.c
index 4b9c7c7..b4d7ff7 100644
--- a/liblog/logd_write.c
+++ b/liblog/logd_write.c
@@ -466,7 +466,11 @@ int __android_log_write(int prio, const char *tag, const char *msg)
         !strcmp(tag, "STK") ||
         !strcmp(tag, "CDMA") ||
         !strcmp(tag, "PHONE") ||
-        !strcmp(tag, "SMS")) {
+        !strcmp(tag, "SMS") ||
+        !strcmp(tag, "FILD") ||
+        !strcmp(tag, "FDL") ||
+        !strcmp(tag, "NVSYNCD") ||
+        !strcmp(tag, "MdmCoreDump")) {
             log_id = LOG_ID_RADIO;
             /* Inform third party apps/ril/radio.. to use Rlog or RLOG */
             snprintf(tmp_tag, sizeof(tmp_tag), "use-Rlog/RLOG-%s", tag);
diff --git a/liblog/logd_write_kern.c b/liblog/logd_write_kern.c
index a7bd61c..473d9ce 100644
--- a/liblog/logd_write_kern.c
+++ b/liblog/logd_write_kern.c
@@ -267,7 +267,11 @@ int __android_log_write(int prio, const char *tag, const char *msg)
         !strcmp(tag, "STK") ||
         !strcmp(tag, "CDMA") ||
         !strcmp(tag, "PHONE") ||
-        !strcmp(tag, "SMS")) {
+        !strcmp(tag, "SMS") ||
+        !strcmp(tag, "FILD") ||
+        !strcmp(tag, "FDL") ||
+        !strcmp(tag, "NVSYNCD") ||
+        !strcmp(tag, "MdmCoreDump")) {
             log_id = LOG_ID_RADIO;
             /* Inform third party apps/ril/radio.. to use Rlog or RLOG */
             snprintf(tmp_tag, sizeof(tmp_tag), "use-Rlog/RLOG-%s", tag);
-- 
2.4.10

